sw/device/tests/sram_ctrl_execution_test_ret.c

```diff
@@ -0,0 +1,120 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include <stdbool.h>
+#include <stdint.h>
+
+#include "sw/device/lib/base/macros.h"
+#include "sw/device/lib/dif/dif_sram_ctrl.h"
+#include "sw/device/lib/handler.h"
+#include "sw/device/lib/runtime/ibex.h"
+#include "sw/device/lib/runtime/log.h"
+#include "sw/device/lib/testing/check.h"
+#include "sw/device/lib/testing/sram_ctrl_testutils.h"
+#include "sw/device/lib/testing/test_framework/test_main.h"
+#include "sw/device/lib/testing/test_framework/test_status.h"
+
+#include "hw/top_earlgrey/sw/autogen/top_earlgrey.h"
+
+const test_config_t kTestConfig;
+
+static dif_sram_ctrl_t sram_ctrl;
+
+/**
+ * This flag is used to verify that the instruction access fault has occurred,
+ * and has been successfully serviced. Declared as volatile, because it is
+ * referenced in the fault handler, as well as the main test flow.
+ */
+static volatile bool instruction_access_fault;
+
+static const uint32_t kRetRamStartAddr =
+    TOP_EARLGREY_SRAM_CTRL_RET_AON_RAM_BASE_ADDR;
+static const uint32_t kRetRamEndAddr =
+    TOP_EARLGREY_SRAM_CTRL_RET_AON_RAM_BASE_ADDR +
+    TOP_EARLGREY_SRAM_CTRL_RET_AON_RAM_SIZE_BYTES - 1;
+
+/**
+ * Handles an exception.
+ *
+ * This exception handler only processes the faults that are relevant. It falls
+ * into an infinite `wait_for_interrupt` routine for the rest.
+ *
+ * The controlled fault originates in the retention SRAM, which means that
+ * normally the return address would be calculated relative to the offending
+ * instruction. However, due to execution from retention SRAM being permanently
+ * disabled, this approach would not work.
+ *
+ * Instead the control needs to be returned to the caller. In other words
+ * sram_execution_test -> retention_sram -> exception_handler
+ * -> sram_execution_test.
+ *
+ * Before the jump into the exception handler, the register set is saved on
+ * stack, which means that the return address can be loaded from there.
+ */
+void handler_exception(void) {
+  uintptr_t ret_addr = (uintptr_t)OT_RETURN_ADDR();
+  LOG_INFO("Fault address: mepc = 0x%x, return address = 0x%x",
+           ibex_mepc_read(), ret_addr);
+
+  exc_id_t exception_id = ibex_mcause_read();
+  switch (exception_id) {
+    case kInstAccFault:
+      LOG_INFO("Instruction access fault handler");
+      instruction_access_fault = true;
+      ibex_mepc_write(ret_addr);
```

Comment:
nit: In the two cases the `ibex_mepc_write`  is equally called, as far as I know there will be no problem to call it to the default case as well. So you could take the call out of the `switch` statement, like after call `ibex_mcause_read()` and reduce one line in the code.

Reply:
That is a fair point, that how I originally had it but then changed as in a default case that would be a redundant operation. In practice this doesn't really matter (just the aesthetics from my perspective). I think for this reason I leave it as it is.